pkg/sources/git/git.go

```diff
@@ -415,7 +416,13 @@ func executeClone(ctx context.Context, params cloneParams) (*git.Repository, err
 		cloneURL.User = params.userInfo
 	}
 
-	gitArgs := []string{"clone", cloneURL.String(), params.clonePath}
+	gitArgs := []string{
+		"clone", cloneURL.String(), params.clonePath,
+		// Fetch additional refs from GitHub and GitLab.
+		// https://github.com/trufflesecurity/trufflehog/issues/1588
+		"-c", "remote.origin.fetch=+refs/pull/*:refs/heads/thog/pr/*",
+		"-c", "remote.origin.fetch=+refs/merge-requests/*:refs/thog/mr/*",
```

Comment:
It occurs to me that `git clone --mirror` may be a better choice. It will automatically map all refs, no need to specify them by hand.

> Set up a mirror of the source repository. This implies --bare. Compared to --bare, --mirror not only maps local branches of the source to local branches of the target, **it maps all refs** (including remote-tracking branches, notes etc.) and sets up a refspec configuration such that all these refs are overwritten by a git remote update in the target repository.
> 
> https://git-scm.com/docs/git-clone#Documentation/git-clone.txt-code--mirrorcode

However, taking this approach would require further refactoring. Just throwing in `--mirror` to the args yields the following error "could not open cloned repo: repository does not exist".

Reply:
What about something like:

```
-c remote.origin.fetch=+refs/*:refs/remotes/origin/*
```
Disclaimer: I'm still playing around with it a bit to make sure it grabs things the way I want it to.

Reply:
That's an option. The benefit of `--mirror` is that it's a bare clone, meaning that it takes up less disk space.

```sh
$ git clone git@github.com:kubernetes/kubernetes.git -c "remote.origin.fetch=+refs/*:refs/remotes/origin/*" k8s-clone-opt
$ du -sh k8s-clone-opt/
2.3G    k8s-clone-opt/

$ git clone git@github.com:kubernetes/kubernetes.git --mirror k8s-mirror
$ du -sh k8s-mirror/
2.1G    k8s-mirror/
```

Reply:
I have a slight preference for `--mirror` assuming there is a comment documenting why we are including `--mirror` in the git clone options.

Reply:
Good point about not checking out a workdir if it's not needed.

I'm not sure if this will impact trufflehog, but I did notice that `--mirror` by itself doesn't seem to map some refs. It pulls all the commits, but I just don't see the branches when running a `git branch -a`, and when I do a gitleaks scan it misses the things on the `replace` branches on GitHub. 

I know this is a trufflehog PR and not a gitleaks one, I'm just using gitleaks here but that's just what I had available and is part of my use case:

Plain mirror: 55 commits scanned:

```
⬢[bplaxco@toolbox tmp.6nM3bbUHs6]$ git clone --mirror https://github.com/leaktk/fake-leaks.git
Cloning into bare repository 'fake-leaks.git'...
remote: Enumerating objects: 290, done.
remote: Counting objects: 100% (178/178), done.
remote: Compressing objects: 100% (112/112), done.
remote: Total 290 (delta 58), reused 164 (delta 46), pack-reused 112
Receiving objects: 100% (290/290), 45.53 KiB | 3.03 MiB/s, done.
Resolving deltas: 100% (81/81), done.
⬢[bplaxco@toolbox tmp.6nM3bbUHs6]$ cd fake-leaks.git/
⬢[bplaxco@toolbox fake-leaks.git]$ gitleaks detect

    ○
    │╲
    │ ○
    ○ ░
    ░    gitleaks

11:03AM INF 55 commits scanned.
11:03AM INF scan completed in 95.4ms
11:03AM WRN leaks found: 43
⬢[bplaxco@toolbox fake-leaks.git]$ git br -a
  2022-10-28-activity-test
  add-license-1
* main
  new-branch
  temp
  test
```

Doing the clone opt one with bare: 82 commits scanned

```
⬢[bplaxco@toolbox tmp.6nM3bbUHs6]$ rm -rf fake-leaks.git/                                                                                                                                                                                                                       
⬢[bplaxco@toolbox tmp.6nM3bbUHs6]$ git clone -c "remote.origin.fetch=+refs/*:refs/remotes/origin/*"  --bare https://github.com/leaktk/fake-leaks.git                                                                                                                           
Cloning into bare repository 'fake-leaks.git'...                                                                                                                                                                                                                                
remote: Enumerating objects: 290, done.                                                                                                                                                                                                                                         
remote: Counting objects: 100% (178/178), done.                                                                                                                                                                                                                                 
remote: Compressing objects: 100% (112/112), done.                                                                                                                                                                                                                              
remote: Total 290 (delta 58), reused 164 (delta 46), pack-reused 112                                                                                                                                                                                                            
Receiving objects: 100% (290/290), 45.53 KiB | 1.20 MiB/s, done.                                                                                                                                                                                                                
Resolving deltas: 100% (81/81), done.                                                                                                                                                                                                                                           
⬢[bplaxco@toolbox tmp.6nM3bbUHs6]$ cd fake-leaks.git/                                                                                                                                                                                                                           
⬢[bplaxco@toolbox fake-leaks.git]$ gitleaks detect                                                                                                                                                                                                                              
                                                                                                                                                                                                                                                                                
    ○                                                                                                                                                                                                                                                                           
    │╲                                                                                                                                                                                                                                                                          
    │ ○                                                                                                                                                                                                                                                                         
    ○ ░                                                                                                                                                                                                                                                                         
    ░    gitleaks                                                                                                                                                                                                                                                               
                                                                                                                                                                                                                                                                                
11:14AM INF 82 commits scanned.                                                                                                                                                                                                                                                 
11:14AM INF scan completed in 152ms                                                                                                                                                                                                                                             
11:14AM WRN leaks found: 62                                                                                                                                                                                                                                                     
⬢[bplaxco@toolbox fake-leaks.git]$ git br -a                                                                                                                                                                                                                                    
  2022-10-28-activity-test
  add-license-1
* main
  new-branch
  temp
  test
  remotes/origin/heads/2022-10-28-activity-test
  remotes/origin/heads/add-license-1
  remotes/origin/heads/main
  remotes/origin/heads/new-branch
  remotes/origin/heads/temp
  remotes/origin/heads/test
  remotes/origin/pull/1/head
  remotes/origin/pull/2/head
  remotes/origin/replace/01ef5781aa6233c6b4141e0cb463b451d9aad0e4
  remotes/origin/replace/07903b4a20c4aef52083668d33e9f3429f61cfe4
  remotes/origin/replace/0b1e0a7bab425242957cca603eac55c5a01631be
  remotes/origin/replace/0d13aa267c8f8e6ed6b8a2e05d9e13f5998435ad
  remotes/origin/replace/0de6e9368155b368c05d0e79c7f72e1119087ea9
  remotes/origin/replace/0e8c13262b5c762a917271ea89ff63c2751c9dd2
  remotes/origin/replace/0f1d68d8e79b3bc324b9af9455c650cd650be766
  remotes/origin/replace/1720153a2d7fe01981b00bb3b2939595c0dce2eb
  remotes/origin/replace/17fe5f788d02ee1b387ab80ee2b7e93e0235987f
  remotes/origin/replace/194eef14cf1264d43a578a57324e5c1fa763e077
...snipped
```

When I used a custom config looking for `auth_token` both cases still found it on that PR branch, but just some hits in the `replace` branches were missing for `--mirror` (even when I gave it custom git fetch config).

I know this PR isn't for gitleaks, but it seemed like the best combo I could get for size and scanning was `--bare -c "remote.origin.fetch=+refs/*:refs/remotes/origin/*"`

It seemed like `--bare` with the config flag was the same size as just `--mirror` by itself but it came with the remote refs in the way that my tooling understood it.

For example here is a commit that it was finding when I did the `--bare` with the fetch opts clone that it wasn't finding in the `--mirror` clone: [01ef5781aa6233c6b4141e0cb463b451d9aad0e4](https://github.com/leaktk/fake-leaks/commit/01ef5781aa6233c6b4141e0cb463b451d9aad0e4). That commit did exist in the `--mirror` clone, just gitleaks wasn't finding it.

Anywho, just wanted to pass that along in case that's helpful, and this is the PR commit that was my original test case:

https://github.com/leaktk/fake-leaks/pull/2/commits/ca8229f5765f2eafa21e07ed780241f62d83f387

Reply:
I think the reason gitleaks was missing it was because of a detail of how git log works though with mirror vs bare with the flags set:

```
⬢[bplaxco@toolbox tmp.6nM3bbUHs6] cd ./bare-with-flags
⬢[bplaxco@toolbox bare-with-flags]$ git log --full-history --all | grep 01ef5781aa6233c6b4141e0cb463b451d9aad0e4                                                                                                                                                                
commit 01ef5781aa6233c6b4141e0cb463b451d9aad0e4                                                                                                                                                                                                                                 
⬢[bplaxco@toolbox bare-with-flags]$ cd ../mirror/                                                                                                                                                                                                                               
⬢[bplaxco@toolbox mirror]$ git log --full-history --all | grep 01ef5781aa6233c6b4141e0cb463b451d9aad0e4
```

Because the commit is for sure in `mirror`:

```
⬢[bplaxco@toolbox mirror]$ git log -n 1 --format=oneline 01ef5781aa6233c6b4141e0cb463b451d9aad0e4 
01ef5781aa6233c6b4141e0cb463b451d9aad0e4 (replaced) Add fake pull secrets
```

But yeah anywho, hopefully trufflehog can get at it with just `--mirror` set because that does feel way cleaner.

Reply:
@bplaxco that's fascinating, I learned something new today. :)

It seems like the current code shows refs from `refs/replace/*`; I'm checking to see how many commits are scanned.

```sh
$ ./trufflehog github --repo="https://github.com/leaktk/fake-leaks.git"
Found unverified result 🐷🔑❓
Detector Type: PrivateKey
Decoder Type: PLAIN
Raw result: -----BEGIN EC PRIVATE KEY-----
MIGkAgEBBDATLrmEm3uwnQaemUAJigVo7ZDYI5LyPWg35b+vXNrOrDpRNxER8lF6
6qm0pZwaPSSgBwYFK4EEACKhZANiAARjEvtRpsFPknM4TR69nkyVnQFNtvl5vtrN
+uiYpD21RdmMriwW8rg+BZ0EpOcDqoMxEvcq6BJfvUVPCuKysoXWjysoplN5GOcK
xa8puL7VGe3zaefbwz1Ajcuf3Okqjxo=
-----END EC PRIVATE KEY-----
Commit: 318a0f1781adf38226c9ade3f6146aeb3e901b18
Commit_source: refs/replace/0de6e9368155b368c05d0e79c7f72e1119087ea9 (hidden ref)
Email: Braxton Plaxco <bplaxco@redhat.com>
File: keys/tls/server-01.key
Line: 1
Link: https://github.com/leaktk/fake-leaks/blob/318a0f1781adf38226c9ade3f6146aeb3e901b18/keys/tls/server-01.key#L1
Repository: https://github.com/leaktk/fake-leaks.git
Timestamp: 2023-05-31 17:50:30 +0000
```